Fixed #28800 -- Added a listurls management command

ulgens opened the pull request:
#### Trac ticket number

ticket-28800

#### Branch description
I added a new management command: `listurls`, which is similar to `showurls` from the django-extensions package.

My initial plan was to push https://github.com/django/django/pull/18347 forward, but reimplementing the ideas from the original command (django-extensions' `show_urls`) was more straightforward. I had to make a couple of design decisions, all of which are open for discussion:

* Command name
  * Current command name is `listurls`.
  * What about `list_urls`? Is it okay to use `_` in a command name?
  * What about `listurlpatterns`? These results originate from `url_patterns` definitions, and the list members are either `RoutePattern` or `RegexPattern`.
  * What about `inspecturls`? Is the action here similar enough to what `inspectdb` does for a database?
* Default format
  * `show_urls`s default format is `dense`. That format looks like a broken table, and it's hard to read, so I didn't copy this format/feature.
  * `aligned` is the new default format. It looks like a table without borders. This one is simple enough to be the default and visually more accessible.
  * `verbose` is more verbose than its django-extensions equivalent.
* Colors
  * Colors from django-extensions looked nice, but I thought it could be better.
  * If there are any accessibility concerns with colors, please let me know.
* Extra features
  * Decorators: django-extensions has `--decorator, -d DECORATOR Show the presence of given decorator on views` param. It may be useful, but it isn't vital for this implementation. I didn't include it.
  * Prefixes: I saw this from https://github.com/django/django/pull/18347, and I agree about its usefulness. It is Included in this PR. The same functionality can be achieved using `grep` for table-based formats, but a lower-level filtering approach was needed for `verbose`, `JSON`, and `pretty-json` formats.

There are another couple of questions I'd like to ask about non-design related topics:
* Even though the work here is not a full or partial import from the django-extensions package, it implements the almost same functionality using the same approach. Does this require special care about licensing?
* The work here is contextually the continuation of https://github.com/django/django/pull/18347 and that PR had the same case too. How do I give credit to @offbyone and @hippietilley?
* I'm not proud of how the tests came out. Please share if you can think of a better way.

**Dense format from django-extensions (not included in the PR)**

![CleanShot 2024-11-14 at 16 16 45@2x](https://github.com/user-attachments/assets/c06eb32b-905b-4489-8679-ee6505afe1d5)

**Aligned (default format)**

![CleanShot 2024-11-13 at 20 47 06@2x](https://github.com/user-attachments/assets/26dd37ee-7151-4227-8691-c0451d59860b)

**Table**

![CleanShot 2024-11-13 at 20 47 55@2x](https://github.com/user-attachments/assets/4873cbf7-b46f-4ba6-b232-16d207847908)

**Verbose**

![CleanShot 2024-11-13 at 20 48 20@2x](https://github.com/user-attachments/assets/e309eae3-29bb-415f-b7e3-5489cc913c27)

Relates to:
- https://github.com/django/django/pull/18347
- https://github.com/hippietilley/django/tree/ticket_28800
- https://forum.djangoproject.com/t/should-the-show-urls-command-be-part-of-djangos-standard-library/20739

#### Checklist
- [x] This PR targets the `main` branch.
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.

ulgens committed:
Fixed #28800 -- Added a listurls management command

ulgens commented:
The tests are not done yet, and I need to update the styling but please let me know if you see a functional issue.

offbyone commented on AUTHORS:
```diff
@@ -1036,6 +1036,7 @@ answer newbie questions, and generally made Django that much better:
     Tyson Clugg <tyson@clugg.net>
     Tyson Tate <tyson@fallingbullets.com>
     Unai Zalakain <unai@gisa-elkartea.org>
+    Ülgen Sarıkavak <https://github.com/ulgens>
```
As I am not already in the authors file, if you want to give me that honor here, I won't turn it down.

ulgens replied:
@offbyone Of course 🌻 Do you have a link you want me to use or is your Github profile okay?

offbyone replied:
Works for me!

shangxiao replied:
Please not in this PR :)

ulgens replied:
@shangxiao Why?

shangxiao replied:
Ah I just saw offbyone was the original PR author, ok should be all good to include in this PR 👍

ulgens replied:
@offbyone Done 🌷

offbyone commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,275 @@
+from django.core.management.base import BaseCommand, CommandError, CommandParser
+from django.utils import termcolors
+
+FORMATS = (
```
You asked about colour; I've seen `--no-color` used as a flag before, maybe use that?

ulgens replied:
It's being used to decide to apply coloring

```python
# Apply colors
if (
    color.supports_color()
    and (not no_color)
    and (format not in COLORLESS_FORMATS)
):
    url_patterns = self.apply_color(url_patterns=url_patterns)
```

offbyone commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,275 @@
+            raise CommandError("There are no URL patterns that match given prefixes")
+
+        unsorted = options["unsorted"]
+        no_color = options["no_color"]
```
I don't see where this is populated.

ulgens replied:
It's coming from `django.core.management.base.BaseCommand`: https://github.com/django/django/blob/4c65aecfe7ebebf2c90f2cccb6d240cd14c1445c/django/core/management/base.py#L350

offbyone reviewed (commented):
Other than a small comment about handling colour, LGTM

shangxiao commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,275 @@
+    def add_arguments(self, parser: CommandParser):
+        super().add_arguments(parser)
+
+        # Sorting
```
Typically "what" comments are discourage 👍

ulgens replied:
I'm not sure what counts as "typical," but in my experience, they provide great value as visual anchors and help decrease the mental load.

shangxiao replied:
I'll tell you what if you find your commenting style _prolifically present in the way you do it in this PR_ anywhere in the Django codebase (there are the occasional what comments) I'll gladly eat my words 🤣 but it's highly likely that you'll be asked to remove them hence "typically". I don't have the final say that's up to the Django fellows...

In my experience what comments are just visual noise that end up diverging and become misleading or just outright incorrect.

But don't take my word for it: https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/

ulgens replied:
@shangxiao Thanks for the details. I'll remove them if the consensus is not to have them, but I don't think that what that blog post describes applies here. Repeating the code in plan English is different from summarizing a code block by a short title. I was going for the second case.

ulgens replied:
Thanks for the opinion.

sarahboyce replied:
> I'll remove them if the consensus is not to have them

Please do not mark comments as resolved if there has been no agreement as it is difficult for other reviewers to reply.

I agree these comments should be removed as the argument names and help text are descriptive enough and we do not add such comments for usages of `parser.add_argument`

ulgens replied:
I removed the comments from the `parser.add_argument()` blocks. I also tried removing the other seemingly redundant comments, but I think the code is better without them. Let me know if there are any other comments you think should be removed.

shangxiao commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,275 @@
+            "-u",
+            action="store_true",
+            dest="unsorted",
+            help="Filter URLs by the given prefix[es]",
```
Help message is wrong here? 🤔

ulgens replied:
Fixed now, thank you.

shangxiao commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,275 @@
+        )
+
+        # Prefix
+        parser.add_argument(
+            "--prefix",
+            "-p",
+            dest="prefixes",
+            help="Only list URLs with these prefixes.",
+            nargs="*",
+        )
```
Just wondering whether `nargs=*` as an option is an anti-pattern or not due to ambiguity? 🤔 

Having a look at all other instances of `nargs=*` in the other commands reveals that they're all regular args. I wonder if this should fall inline with the convention? 🤔 

Though the Python docs mention it's fine to do so 🤷 https://docs.python.org/3/library/argparse.html#nargs

ulgens replied:
I don't think copying the exact same approach from other commands is always a good thing. I needed a way to accept multiple "prefix" arguments and `nargs="*"` does that.

shangxiao commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,275 @@
+
+    def handle(self, *args, **options):
+        # Make sure the prefixes are a list
+        prefixes = options["prefixes"]
+        if prefixes:
+            prefixes = prefixes if isinstance(prefixes, list) else [prefixes]
```
nargs of * is always a list so it's likely you'll be asked to remove the check, (the other commands don't check for types of list as well)

ulgens replied:
I remember a weird case happening here, but I tested it again and couldn't replicate it. Removing the check and the conversion, thank you.

shangxiao commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,275 @@
+            view_str = f"{self.style.HEADER('View:')} {view}"
+            name_str = f"{self.style.HEADER('Name:')} {name}" if name else ""
+
+            seperator = "-" * 20 + "\n"
```
spelling: `separator`

shangxiao commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,275 @@
+        indent = 4 if pretty else None
+
+        # Having keys in the resulting JSON makes it more useful
+        url_pattern_dicts = []
+        for route, view, name in url_patterns:
+            url_pattern_dict = {
+                "route": route,
+                "view": view,
+                "name": name,
+            }
+            url_pattern_dicts.append(url_pattern_dict)
```
Comprehension? 🤔

ulgens replied:
Makes sense, thank you.

shangxiao commented on docs/ref/django-admin.txt:
```diff
@@ -498,6 +498,18 @@ Only support for PostgreSQL is implemented.
+``listurls``
+------------
+
+.. django-admin:: listurls [url_prefix ...]
```
Doesn't match API 😅

ulgens replied:
Fixed now.

shangxiao reviewed (commented):
Some prelim feedback ☺️

ulgens committed:
Fix typo in word "separator"

ulgens committed:
Use list comprehension instead a for loop

ulgens committed:
Update help message to match the related argument

smithdc1 commented:
Thank you for the work here. 

I think we should take the proposed formats (I make it 5?) to the forum thread to try and gain concensus on two questions. 

- Are we happy with the 5 options? If less then which ones? If more, what?

- For the formats agreed upon are there any specific comments? (May be a tweak here or there ).

I'm happy to help write that up if it would be useful.

ulgens commented:
@smithdc1 Thanks for your notes. While moving that to the forum, is it okay to continue from the existing one, or does this deserve another forum post?

I could use some help there; otherwise, I am not sure I can handle it before the weekend.

ulgens committed:
Remove unnecessary type check - conversion

ulgens committed:
Add Chris Rose (offbyone) to AUTHORS file

ulgens committed:
Update the command explanation in docs

ulgens committed:
Handle no_color flag in formats with colored titles

ulgens commented:
@smithdc1 

> I noticed that passing these options result in a crash.

This should be okay now, thanks for letting me know. `--force-color` causes the same issue too but I will be able to check it another day.

One question, though: What should be the testing approach when the output and test case are almost exactly the same but there is only a small difference between them, like the difference `--no-color` makes? Normally, I would use `parameterized`, but it means an extra dependency. I can create `for` loops in test cases or just duplicate them for each combination, but I'm not sure they are the better options.

bmispelon commented:
Amazing work, I'm excited to see this land in Django! 🚀 

You mention that this was inspired by django-extensions' `show_urls`:
1) I believe [their license](https://github.com/django-extensions/django-extensions/blob/main/LICENSE) means we need to acknowledge their copyright in our source code (I'm no expert, but I think a not at the top of the `listurls.py` file might be enough)
2) I think we should include a link to the original project at least in the release notes and maybe even on the command's documentation page. We're building on top of their work and that should be acknowledged publicly.

rtpg commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,277 @@
+
+    def handle(self, *args, **options):
+        prefixes = options["prefixes"]
+        url_patterns = get_url_patterns(prefixes=prefixes)
```
Can we make this call happen in a class method? So people could override this `Command` with their own variant that might filter down URL patterns based on some property

ulgens replied:
Thanks for catching this. It makes perfect sense, I'll update the code.

ulgens replied:
@rtpg This should be okay now. Please let me know if the new version is not up to what you had in mind.

shangxiao commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,277 @@
+        if self.is_color_enabled:
+            headers = [self.style.HEADER(header) for header in headers]
+
+        header_parts = []
+        for width, header in zip(widths, headers, strict=False):
+            header_parts.append(header.ljust(width))
```
Oh this is also comp potential 🕵️‍♂️

ulgens replied:
The equivalent of this code from django-extensions was using comprehension, I chose this version for better readability.

pauloxnet replied:
The comprehension version can perform better, and at the same time be readable as it's a common Python pattern.

shangxiao commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,277 @@
+        formatted_str.write("\n")
+
+        # Header - content separator
+        formatted_str.write("-+-".join("-" * width for width in widths))
```
Also you'll likely be asked to wrap the genexp in a list comprehension when doing a `join()` as it's more performant…  I don't have the link handy but felix knows what I'm talking about (he's the one who asks for it).

Having said that, this is a command & not a hot path so probably doesn't matter and there may be a preference for readability.

ulgens replied:
Readability does matter but I couldn't catch what you meant. Instead of referring to other people and their thoughts, it would be nicer if you can just add a suggestion.

shangxiao replied:
@ulgens Can you keep your tone friendly & respectful please?  I get that it can be hard to have other people review your code as it feels like someone just loading on the criticisms… I've been trying hard to make my feedback as friendly as possible but the responses I feel like I'm getting from you are just a tad too "I know better than you" type of tone 🤷

If you like we can jump on Discord to resolve this amicably 👍 ☺️

ulgens replied:
@shangxiao I'm not sure where this is coming from but I assure you, there is no "I know better than you" attitude in my response. 

Again, if you can explain your comment better by giving an example, that would be nice of you. Otherwise I'm not sure how to act on a comment that I don't understand.

sarahboyce replied:
I can understand why @shangxiao is frustrated.
Some of these comments are dismissive and resolving comments before an agreement is also dismissive.
This feels disrespectful to the time that folks give to review a PR. 

Also note that @shangxiao  is a member of the triage and review team and an active contributor to Django - these are not fly-by comments but based off experience contributing to this project.

Part of the reason a suggestion hasn't been given here is it's not clear if we should make this update or not.
"felix" is @felixxm and I would like him to weigh in here :+1:

ulgens replied:
@shangxiao Almost none of us are paid members, so I believe we all agree that being nice is the key 🍻  That being said, I was only asking for an actionable insight and I'm still surprised how it came to this, but I'll be more careful while responding to incoming comments.

About the list comprehension comment, thanks for the example. I'm still missing the benefit it provides, but if that's some styling/linting expectation, I'll apply that.

nessita replied:
For further reference, I believe this is what @shangxiao refers to:
https://github.com/django/django/pull/16873#issuecomment-1554421115

Copying from that comment:
> list comprehension is preferable here as str.join() converts to list internally anyway. It is better performance to provide a list up front.

https://stackoverflow.com/questions/9060653/list-comprehension-without-in-python/9061024#9061024

adamchainz/flake8-comprehensions#156

ulgens replied:
@nessita Thanks for the resources. It seems the overall discussion wasn't finalized and there is open question about the tradeoff it requires: https://github.com/adamchainz/flake8-comprehensions/issues/156#issuecomment-2459089173 

In either case, I don't think the performance difference matters enough to have a detailed discussion about this change:

```bash
> python -m timeit '"".join(str(n) for n in range(20))'
200000 loops, best of 5: 1.56 usec per loop

> python -m timeit '"".join([str(n) for n in range(20)])'
200000 loops, best of 5: 1.19 usec per loop
```

The difference between these two is 0.4 milliseconds.

I applied the list comprehension to two different places this case applies. I hope that's okay with everyone.

ulgens committed:
Make get_url_patterns into a class method

ulgens committed:
Move list to dict conversion to it's own method, from .format_json()

ulgens commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,277 @@
+        # Having keys in the resulting JSON makes it more useful
+        url_pattern_dicts = [
+            {"route": route, "view": view, "name": name}
+            for route, view, name in url_patterns
```
I moved the dict conversion to new `as_dict` method. Can you please check the final result? Thank you.

ulgens committed:
Wrap generator expression in join() with list comprehension for a slight performance increase

https://stackoverflow.com/a/9061024/1726238
https://github.com/adamchainz/flake8-comprehensions/issues/156#issuecomment-2459089173

ulgens committed:
Remove the redundant comments, the code itself is already clear enough

sarahboyce commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,277 @@
+        colored_url_patterns = []
+
+        for url_pattern in url_patterns:
+            # Route
+            route = url_pattern[0]
```
I think using a namedtuple (or a different data structure) will give us the readability we want here
Something like
```python
URLPattern = namedtuple("URLPattern", ["route", "view", "name"])
```

Then this section would become:
```python
            route = self.style.ROUTE(url_pattern.route)

            module_path, module_name = url_pattern.view.rsplit(".", 1)
            module_name = self.style.VIEW_NAME(module_name)
            view = f"{module_path}.{module_name}"

            name = url_pattern.name
            if name:
                namespace, name = name.rsplit(":", 1) if ":" in name else ("", name)
                name = self.style.NAME(name)
                name = f"{namespace}:{name}" if namespace else name

            url_pattern = URLPattern(route=route, view=view, name=name)
            url_patterns.append(url_pattern)
```

We can become any index reference and it has another advantage that we can remove `as_dict` as this is inbuilt:
```diff
         return formatted_str.getvalue()
 
-    def as_dict(self, url_pattern):
-        return {
-            "route": url_pattern[0],
-            "view": url_pattern[1],
-            "name": url_pattern[2],
-        }
-
     def format_json(self, url_patterns, pretty=False):
         indent = 4 if pretty else None
 
-        url_pattern_dicts = [self.as_dict(url_pattern) for url_pattern in url_patterns]
+        url_pattern_dicts = [url_pattern._asdict() for url_pattern in url_patterns]
```

sarahboyce commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,277 @@
+
+        self.style.HEADER = termcolors.make_style(opts=("bold",))
+        self.style.ROUTE = termcolors.make_style(opts=("bold",))
+        self.style.VIEW_NAME = termcolors.make_style(fg="yellow", opts=("bold",))
+        self.style.NAME = termcolors.make_style(fg="red", opts=("bold",))
```
```suggestion
        self.style.HIGHLIGHT = termcolors.make_style(fg="yellow", opts=("bold",))
```
I think red is associated with errors and so I would change this to be a highlight style.
Part of the reason I think the different colors existed before was around the visual difference due to the dense format, without that format I think 1 color is enough

sarahboyce commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,277 @@
+from django.core.management.base import BaseCommand, CommandError, CommandParser
+from django.utils import termcolors
+
+FORMATS = (
+    "aligned",
+    "table",
+    "verbose",
+    "json",
+    "pretty-json",
+)
```
I have added a comment here on the formats we are supporting: https://forum.djangoproject.com/t/should-the-show-urls-command-be-part-of-djangos-standard-library/20739/16 

The main thing would be having 1 format be approved as accessible from the accessibility team.

sarahboyce commented on docs/ref/django-admin.txt:
```diff
@@ -498,6 +498,26 @@ Only support for PostgreSQL is implemented.
 
+``listurls``
+------------
+
```
```suggestion

.. versionadded:: 5.2

```

sarahboyce commented on django/core/management/commands/listurls.py:
```diff
@@ -0,0 +1,277 @@
+        )
+
+        if self.is_color_enabled:
+            url_patterns = self.apply_color(url_patterns=url_patterns)
```
This is missing test coverage (and the `apply_color` method)

sarahboyce reviewed (commented):
Thank you @ulgens - I have initial comments :star:

nessita commented:
Closed in favor of #21307.
